metrics/stackdriver_exporter.go

```diff
@@ -113,32 +159,92 @@ func newOpencensusSDExporter(o stackdriver.Options) (view.Exporter, error) {
+	if err != nil {
+		logger.Errorw("Failed to create the Stackdriver exporter: ", zap.Error(err))
+		return nil, nil, err
+	}
+	logger.Infof("Created Opencensus Stackdriver exporter with config %v", config)
+	// We have to return a ResourceExporterFactory here to enable tracking resources, even though we always poll for them.
+	return &pollOnlySDExporter{e},
+		func(r *resource.Resource) (view.Exporter, error) { return &pollOnlySDExporter{}, nil },
+		nil
+}
+
+func generateStackdriverOptions(config *metricsConfig, logger *zap.SugaredLogger) stackdriver.Options {
 	gm := getMergedGCPMetadata(config)
 	mpf := getMetricPrefixFunc(config.stackdriverMetricTypePrefix, config.stackdriverCustomMetricTypePrefix)
 	co, err := getStackdriverExporterClientOptions(config)
 	if err != nil {
 		logger.Warnw("Issue configuring Stackdriver exporter client options, no additional client options will be used: ", zap.Error(err))
 	}
-	// Automatically fall back on Google application default credentials
-	e, err := newStackdriverExporterFunc(stackdriver.Options{
+
+	return stackdriver.Options{
 		ProjectID:               gm.project,
 		Location:                gm.location,
 		MonitoringClientOptions: co,
 		TraceClientOptions:      co,
 		GetMetricPrefix:         mpf,
-		ResourceByDescriptor:    getResourceByDescriptorFunc(config.stackdriverMetricTypePrefix, gm),
 		ReportingInterval:       config.reportingPeriod,
 		DefaultMonitoringLabels: &stackdriver.Labels{},
-	})
-	if err != nil {
-		logger.Errorw("Failed to create the Stackdriver exporter: ", zap.Error(err))
-		return nil, err
 	}
-	logger.Infof("Created Opencensus Stackdriver exporter with config %v", config)
-	return e, nil
+}
+
+func sdCustomMetricsRecorder(mc metricsConfig) func(context.Context, []stats.Measurement, ...stats.Options) error {
+	return func(ctx context.Context, mss []stats.Measurement, ros ...stats.Options) error {
+		// Filter the measuremenst array to only include permitted metrics.
+		i := 0
+		var templ *resourceTemplate
+		for _, m := range mss {
+			metricType := path.Join(mc.stackdriverMetricTypePrefix, m.Measure().Name())
+			if t, ok := metricToResourceLabels[metricType]; ok {
+				mss[i] = m
+				i++
+				if templ != nil && templ != t {
+					return fmt.Errorf("Mixed resource type measurements in one report: %v", mss)
+				}
+				templ = t
+			}
+		}
+		// Trim the array
+		mss = mss[:i]
+		if i == 0 {
+			return nil
+		}
+		// Extract resource, if possible
+		if templ != nil {
+			tagMap := tag.FromContext(ctx)
+			r := resource.Resource{
+				Type:   templ.Type,
+				Labels: map[string]string{},
+			}
+			tagMutations := make([]tag.Mutator, 0, len(templ.LabelKeys))
+			for k := range templ.LabelKeys {
+				tagKey := tag.MustNewKey(k)
+				if v, ok := tagMap.Value(tagKey); ok {
+					r.Labels[k] = v
+					tagMutations = append(tagMutations, tag.Delete(tagKey))
+				}
+			}
+			var err error
+			ctx, err = tag.New(metricskey.WithResource(ctx, r), tagMutations...)
```

Comment:
I don't understand how we deal with resourceTemplate.

`getResourceByDescriptorFunc` is removed from `stackdriver.Option` in line 132 of original file. It looks like we depends on the resource in context to decide which monitored resource type (concept of stackdriver) and resource type labels to use for a metric. I can see the resource is built into context here, and consumed by `stats.RecordWithOptions`. However I don't get how this information is populated to a Stackdriver exporter.

Reply:
(I removed all of that and rewrote bits of `stackdriver_exporter_test.go` to use metrics export instead.)

Reply:
Thanks for explaining. `newStackdriverExporter` returns an exporter, an exporter factory func. Let's call the returned exporter from `newStackdriverExporter` `E1` and the one created by the returned factory func `E2`.

 IIUC, `E1` is a real SD exporter and is actually running. `E2` is a `pollOnlySDExporter` without a real internal SD exporter. `mE.m` sets the resource, it also registers `E2` via `mE.m.RegisterExporter` instead of `E1`. Does `E1` also knows the Resource associated with the given Meter?

Reply:
I think this may have changed thanks to some work by @jjzeng-seattle and I in OpenCensus: https://github.com/census-instrumentation/opencensus-go/pull/1212

In particular, the Stackdriver implementation prefers to use its own internal timer and batching, rather than being pushed metrics via the Exporter API. See https://github.com/census-ecosystem/opencensus-go-exporter-stackdriver/blob/master/stackdriver.go#L404 for details.

What happens now is that Stackdriver uses the same path as Prometheus, and we never need to (and never should) call `view.RegisterExporter(E*)`. Instead we call `e.StartMetricsExporter()` on line 166.

Reply:
yes.  It's pretty messy for opencensus here.

Prometheus exporter is "pull" mode only, when means it collects metrics when the scrapers call.  It's a no-op when the metrics are pushed to it view `ExportView` https://github.com/census-ecosystem/opencensus-go-exporter-prometheus/blob/master/prometheus.go#L109

Opencensus exporter is "push" mode only, and it doesn't have an internal reader to collect metrics.  It gets metrics from `ExportView`.  https://github.com/census-ecosystem/opencensus-go-exporter-prometheus/blob/master/prometheus.go#L109

Stackdriver exporter is first implemented in "push" mode, then switched to "pull" mode.  You cannot enable both modes, and the "pull" mode is recommended.  The comments here explain it pretty clear: https://github.com/census-ecosystem/opencensus-go-exporter-prometheus/blob/master/prometheus.go#L109

So here we use `E2` to wrap `E1` so we won't call `RegisterExporter(E1)`, which introduces undefined behavior.

Reply:
I'm hoping that in O(3-4 releases) we can simplify this code down to a single export (Opencensus), and then we don't need to discuss push vs pull, but the current state of the world is as JJ describes.